lib/gooddata/lcm/lcm2.rb

```diff
@@ -15,38 +15,35 @@ module LCM2
     class SmartHash < Hash
       @specification = nil
       def method_missing(name, *_args)
-        key = name.to_s.downcase.to_sym
-
-        value = nil
-        keys.each do |k|
-          if k.to_s.downcase.to_sym == key
-            value = self[k]
-            break
-          end
-        end
-
-        if value
-          value
-        else
-          begin
-            super
-          rescue
-            nil
-          end
-        end
+        data(name)
       end
 
       def [](variable)
-        fail "Param #{variable} is not defined in the specification" if @specification && !@specification[variable] && !@specification[variable.to_sym]
-        super(variable)
+        data(variable)
       end
 
       def clear_filters
         @specification = nil
       end
 
       def setup_filters(filter)
-        @specification = filter
+        @specification = filter.to_hash
+      end
+
+      def check_specification(variable)
+        if @specification && !@specification[variable.to_sym] && !@specification[variable.to_s] \
+                          && !@specification[variable.to_s.downcase.to_sym] && !@specification[variable.to_s.downcase]
+          fail "Param #{variable} is not defined in the specification"
+        end
+      end
+
+      def data(variable)
+        check_specification(variable)
+        final_value = nil
+        keys.each do |k|
+          final_value = values_at(k).first if k.to_s.downcase.to_sym == variable.to_s.downcase.to_sym
+        end
```

Comment:
Would it be better like this? `keys.find { |k|  k.to_s.downcase.to_sym == variable.to_s.downcase.to_sym}`